### ISSUE TITLE
publish() and subscribe() to a topic are not executed linear/sync

### ISSUE THREAD

**Issue by mikeandtherest** [uNetworking/uWebSockets.js#301]
Hi guys,

I ran into this weird issue with ws.publish() and ws.subscribe()  seen in the sample below:

```javascript
  message: (ws, message, isBinary) => {
    /* first I do the publish for all the current subcribers */
    ws.publish('smth', payload, true);

    /* and only after that I subscribe the current socket to the same topic */
    ws.subscribe('smth');
  }
```

But it seems that the current socket also receives the notification. I tried it also with `this.wsApp.publish(...)` and it has the same result.

The only thing that fixes the problem, is when I do the subscription on the next cycle, by having a 0 setTimeout(), like:

```javascript
  message: (ws, message, isBinary) => {
    ws.publish('smth', payload, true);

    setTimeout(() => {
        ws.subscribe('smth');
    }
```

Now I don't know if there is something that I do wrong or if it's an actual issue with the lib.

But let's say that I'd stick to this workaround. Could it imply any kind of risk/issue or is it OK?

Thanks.

**Comment by alexhultman** [comment]
That workaround is fine for now, but I think you're looking for the feature to publish to all except for the sender? That's an upcoming feature I'm working on.

Subscribe/unsubscribe is immediate, but publish happens when the loop iteration ends, so that's why your workaround works fine for now.

**Comment by mikeandtherest** [comment]
Yes, that's what I'm trying to achieve, i.e. to publish to all except for the current ws. 
Wasn't aware that this is something that will come later, and that's why my first hunch was that it should work with  a publish followed by a subscribe. But I  understand now what the actual flow is, thx for the details.

Cool, if the setTimeout workaround doesn't add any kind of side effects, that's all I need for now.

Should we keep this open for reference and label it to something, or I'd better close it?

**Comment by alexhultman-2fa** [comment]
This is fixed in 18.2; it will be strict there
